consensus/src/pending_votes.rs

```diff
@@ -56,15 +62,120 @@ pub enum VoteReceptionResult {
     EchoTimeout(u128),
 }
 
+#[derive(Debug)]
+pub(super) struct TwoChainTimeoutVotes {
+    timeout_reason: HashMap<Author, RoundTimeoutReason>,
+    partial_2chain_tc: TwoChainTimeoutWithPartialSignatures,
+}
+
+impl TwoChainTimeoutVotes {
+    fn new(timeout: TwoChainTimeout) -> Self {
+        Self {
+            partial_2chain_tc: TwoChainTimeoutWithPartialSignatures::new(timeout.clone()),
+            timeout_reason: HashMap::new(),
+        }
+    }
+
+    fn add(
+        &mut self,
+        author: Author,
+        timeout: TwoChainTimeout,
+        signature: bls12381::Signature,
+        reason: RoundTimeoutReason,
+    ) {
+        self.partial_2chain_tc.add(author, timeout, signature);
+        self.timeout_reason.entry(author).or_insert(reason);
+    }
+
+    pub(super) fn partial_2chain_tc_mut(&mut self) -> &mut TwoChainTimeoutWithPartialSignatures {
+        &mut self.partial_2chain_tc
+    }
+
+    pub(super) fn partial_2chain_tc(&self) -> &TwoChainTimeoutWithPartialSignatures {
+        &self.partial_2chain_tc
+    }
+
+    fn aggregated_timeout_reason(&self, verifier: &ValidatorVerifier) -> RoundTimeoutReason {
+        let mut reason_voting_power: HashMap<RoundTimeoutReason, u128> = HashMap::new();
+        let mut missing_batch_authors: HashMap<Author, u128> = HashMap::new();
+        let ordered_authors = verifier.get_ordered_account_addresses();
+        for (author, reason) in &self.timeout_reason {
+            // To aggregate the reason, we only care about the variant type itself and
+            // exclude any data within the variants.
+            let reason_key = match reason {
+                reason @ RoundTimeoutReason::Unknown
+                | reason @ RoundTimeoutReason::ProposalNotReceived
+                | reason @ RoundTimeoutReason::NoQC => reason.clone(),
+                RoundTimeoutReason::PayloadUnavailable { missing_authors } => {
+                    for missing_idx in missing_authors.iter_ones() {
+                        if let Some(missing_author) = ordered_authors.get(missing_idx) {
+                            *missing_batch_authors.entry(*missing_author).or_default() +=
+                                verifier.get_voting_power(missing_author).unwrap() as u128;
```

Comment:
i suppose you mean the voting power of the author instead of the missing_author?